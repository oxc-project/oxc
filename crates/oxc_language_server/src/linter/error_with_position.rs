use std::{path::PathBuf, str::FromStr};

use tower_lsp::lsp_types::{
    self, CodeDescription, DiagnosticRelatedInformation, NumberOrString, Position, Range, Url,
};

use cow_utils::CowUtils;
use oxc_diagnostics::{Error, Severity};

use crate::linter::offset_to_position;

const LINT_DOC_LINK_PREFIX: &str = "https://oxc.rs/docs/guide/usage/linter/rules";

#[derive(Debug)]
pub struct ErrorWithPosition {
    pub start_pos: Position,
    pub end_pos: Position,
    pub miette_err: Error,
    pub fixed_content: Option<FixedContent>,
    pub labels_with_pos: Vec<LabeledSpanWithPosition>,
}

#[derive(Debug)]
pub struct LabeledSpanWithPosition {
    pub start_pos: Position,
    pub end_pos: Position,
    pub message: Option<String>,
}

#[derive(Debug, Clone)]
pub struct DiagnosticReport {
    pub diagnostic: lsp_types::Diagnostic,
    pub fixed_content: Option<FixedContent>,
}
#[derive(Debug)]
pub struct ErrorReport {
    pub error: Error,
    pub fixed_content: Option<FixedContent>,
}

#[derive(Debug, Clone)]
pub struct FixedContent {
    pub code: String,
    pub range: Range,
}

fn cmp_range(first: &Range, other: &Range) -> std::cmp::Ordering {
    match first.start.cmp(&other.start) {
        std::cmp::Ordering::Equal => first.end.cmp(&other.end),
        o => o,
    }
}

/// parse `OxcCode` to `Option<(scope, number)>`
fn parse_diagnostic_code(code: &str) -> Option<(&str, &str)> {
    if !code.ends_with(')') {
        return None;
    }
    let right_parenthesis_pos = code.rfind('(')?;
    Some((&code[0..right_parenthesis_pos], &code[right_parenthesis_pos + 1..code.len() - 1]))
}

impl ErrorWithPosition {
    pub fn new(
        error: Error,
        text: &str,
        fixed_content: Option<FixedContent>,
        start: usize,
    ) -> Self {
        let labels = error.labels().map_or(vec![], Iterator::collect);
        let labels_with_pos: Vec<LabeledSpanWithPosition> = labels
            .iter()
            .map(|labeled_span| LabeledSpanWithPosition {
                start_pos: offset_to_position(labeled_span.offset() + start, text),
                end_pos: offset_to_position(
                    labeled_span.offset() + start + labeled_span.len(),
                    text,
                ),
                message: labeled_span.label().map(ToString::to_string),
            })
            .collect();

        let start_pos = labels_with_pos[0].start_pos;
        let end_pos = labels_with_pos[labels_with_pos.len() - 1].end_pos;

        Self { miette_err: error, start_pos, end_pos, labels_with_pos, fixed_content }
    }

    fn to_lsp_diagnostic(&self, path: &PathBuf) -> lsp_types::Diagnostic {
        let severity = match self.miette_err.severity() {
            Some(Severity::Error) => Some(lsp_types::DiagnosticSeverity::ERROR),
            _ => Some(lsp_types::DiagnosticSeverity::WARNING),
        };
        let related_information = Some(
            self.labels_with_pos
                .iter()
                .map(|labeled_span| lsp_types::DiagnosticRelatedInformation {
                    location: lsp_types::Location {
                        uri: lsp_types::Url::from_file_path(path).unwrap(),
                        range: lsp_types::Range {
                            start: lsp_types::Position {
                                line: labeled_span.start_pos.line,
                                character: labeled_span.start_pos.character,
                            },
                            end: lsp_types::Position {
                                line: labeled_span.end_pos.line,
                                character: labeled_span.end_pos.character,
                            },
                        },
                    },
                    message: labeled_span.message.clone().unwrap_or_default(),
                })
                .collect(),
        );
        let range = related_information.as_ref().map_or(
            Range { start: self.start_pos, end: self.end_pos },
            |infos: &Vec<DiagnosticRelatedInformation>| {
                let mut ret_range = Range {
                    start: Position { line: u32::MAX, character: u32::MAX },
                    end: Position { line: u32::MAX, character: u32::MAX },
                };
                for info in infos {
                    if cmp_range(&ret_range, &info.location.range) == std::cmp::Ordering::Greater {
                        ret_range = info.location.range;
                    }
                }
                ret_range
            },
        );
        let code = self.miette_err.code().map(|item| item.to_string());
        let code_description = code.as_ref().and_then(|code| {
            let (scope, number) = parse_diagnostic_code(code)?;
            Some(CodeDescription {
                href: Url::from_str(&format!(
                    "{LINT_DOC_LINK_PREFIX}/{}/{number}",
                    scope.strip_prefix("eslint-plugin-").unwrap_or(scope).cow_replace("-", "_")
                ))
                .ok()?,
            })
        });
        let message = self.miette_err.help().map_or_else(
            || self.miette_err.to_string(),
            |help| format!("{}\nhelp: {}", self.miette_err, help),
        );

        lsp_types::Diagnostic {
            range,
            severity,
            code: code.map(NumberOrString::String),
            message,
            source: Some("oxc".into()),
            code_description,
            related_information,
            tags: None,
            data: None,
        }
    }

    pub fn into_diagnostic_report(self, path: &PathBuf) -> DiagnosticReport {
        DiagnosticReport {
            diagnostic: self.to_lsp_diagnostic(path),
            fixed_content: self.fixed_content,
        }
    }
}
