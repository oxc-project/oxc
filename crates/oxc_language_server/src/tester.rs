use std::{fmt::Write, path::PathBuf};

use tower_lsp_server::{
    UriExt,
    lsp_types::{CodeDescription, NumberOrString, Uri},
};

use crate::{Options, worker::WorkspaceWorker};

use super::linter::error_with_position::DiagnosticReport;

/// Given a file path relative to the crate root directory, return the absolute path of the file.
pub fn get_file_path(relative_file_path: &str) -> PathBuf {
    std::env::current_dir().expect("could not get current dir").join(relative_file_path)
}

/// Given a file path relative to the crate root directory, return the URI of the file.
pub fn get_file_uri(relative_file_path: &str) -> Uri {
    Uri::from_file_path(get_file_path(relative_file_path))
        .expect("failed to convert file path to URL")
}

fn get_snapshot_from_report(report: &DiagnosticReport) -> String {
    let code = match &report.diagnostic.code {
        Some(NumberOrString::Number(code)) => code.to_string(),
        Some(NumberOrString::String(code)) => code.to_string(),
        None => "None".to_string(),
    };
    let code_description_href = match &report.diagnostic.code_description {
        Some(CodeDescription { href }) => href.to_string(),
        None => "None".to_string(),
    };
    let message = report.diagnostic.message.clone();
    let range = report.diagnostic.range;
    let related_information = match &report.diagnostic.related_information {
        Some(infos) => {
            infos
                .iter()
                .enumerate()
                .map(|(i, info)| {
                    let mut result = String::new();
                    write!(result, "related_information[{}].message: {:?}", i, info.message)
                        .unwrap();
                    // replace everything between `file://` and `oxc_language_server` with `<variable>`, to avoid
                    // the absolute path causing snapshot test failures in different environments
                    let mut location = info.location.uri.to_string();
                    let start =
                        location.find("file://").expect("file:// protocol not found in URI");
                    let end = location
                        .find("oxc_language_server")
                        .expect("oxc_language_server not found in URI");
                    location.replace_range(
                        start + "file://".len()..end + "oxc_language_server".len(),
                        "<variable>",
                    );

                    write!(result, "\nrelated_information[{i}].location.uri: {location:?}",)
                        .unwrap();
                    write!(
                        result,
                        "\nrelated_information[{}].location.range: {:?}",
                        i, info.location.range
                    )
                    .unwrap();
                    result
                })
                .collect::<Vec<_>>()
                .join("\n")
        }
        None => "related_information: None".to_string(),
    };
    let severity = report.diagnostic.severity;
    let source = &report.diagnostic.source;
    let tags = &report.diagnostic.tags;
    let fixed = &report.fixed_content;

    format!(
        r"
code: {code:?}
code_description.href: {code_description_href:?}
message: {message:?}
range: {range:?}
{related_information}
severity: {severity:?}
source: {source:?}
tags: {tags:?}
fixed: {fixed:?}
"
    )
}

/// Testing struct for the [linter server][crate::linter::server_linter::ServerLinter].
pub struct Tester<'t> {
    relative_root_dir: &'t str,
    options: Option<Options>,
}

impl Tester<'_> {
    pub fn new(relative_root_dir: &'static str, options: Option<Options>) -> Self {
        Self { relative_root_dir, options }
    }

    async fn create_workspace_worker(&self) -> WorkspaceWorker {
        let absolute_path = std::env::current_dir()
            .expect("could not get current dir")
            .join(self.relative_root_dir);
        let uri = Uri::from_file_path(absolute_path).expect("could not convert current dir to uri");
        let worker = WorkspaceWorker::new(uri);
        worker.init_linter(&self.options.clone().unwrap_or_default()).await;

        worker
    }

    /// Given a relative file path (relative to `oxc_language_server` crate root), run the linter
    /// and return the resulting diagnostics in a custom snapshot format.
    #[expect(clippy::disallowed_methods)]
    pub fn test_and_snapshot_single_file(&self, relative_file_path: &str) {
        let uri = get_file_uri(&format!("{}/{}", self.relative_root_dir, relative_file_path));
        let reports = tokio::runtime::Runtime::new().unwrap().block_on(async {
            self.create_workspace_worker()
                .await
                .lint_file(&uri, None)
                .await
                .expect("lint file is ignored")
        });
        let snapshot = if reports.is_empty() {
            "No diagnostic reports".to_string()
        } else {
            reports.iter().map(get_snapshot_from_report).collect::<Vec<_>>().join("\n")
        };

        let snapshot_name = self.relative_root_dir.replace('/', "_");
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_omit_expression(true);
        if let Some(path) = uri.to_file_path() {
            settings.set_input_file(path.as_ref());
        }
        settings.set_snapshot_suffix(relative_file_path.replace('/', "_"));
        settings.bind(|| {
            insta::assert_snapshot!(snapshot_name, snapshot);
        });
    }
}
