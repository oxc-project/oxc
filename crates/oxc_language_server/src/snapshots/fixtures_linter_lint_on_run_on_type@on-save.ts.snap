---
source: crates/oxc_language_server/src/tester.rs
---
########## 
file: fixtures/linter/lint_on_run/on_type/on-save.ts
----------

code: "eslint(no-debugger)"
code_description.href: "https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-debugger.html"
message: "`debugger` statement is not allowed\nhelp: Remove the debugger statement"
range: Range { start: Position { line: 1, character: 0 }, end: Position { line: 1, character: 9 } }
related_information[0].message: ""
related_information[0].location.uri: "file://<variable>/fixtures/linter/lint_on_run/on_type/on-save.ts"
related_information[0].location.range: Range { start: Position { line: 1, character: 0 }, end: Position { line: 1, character: 9 } }
severity: Some(Warning)
source: Some("oxc")
tags: None
fixed: Multiple(
    [
        FixedContent {
            message: Some(
                "Remove the debugger statement",
            ),
            code: "",
            range: Range {
                start: Position {
                    line: 1,
                    character: 0,
                },
                end: Position {
                    line: 1,
                    character: 9,
                },
            },
        },
        FixedContent {
            message: Some(
                "Disable no-debugger for this line",
            ),
            code: "// oxlint-disable-next-line no-debugger\n",
            range: Range {
                start: Position {
                    line: 1,
                    character: 0,
                },
                end: Position {
                    line: 1,
                    character: 0,
                },
            },
        },
        FixedContent {
            message: Some(
                "Disable no-debugger for this file",
            ),
            code: "// oxlint-disable no-debugger\n",
            range: Range {
                start: Position {
                    line: 0,
                    character: 0,
                },
                end: Position {
                    line: 0,
                    character: 0,
                },
            },
        },
    ],
)
